### PR TITLE
Boost: Add toast notification when toggling cache logging

### PR DESCRIPTION
**Pull request by dilirity**
Part of https://github.com/Automattic/jetpack/issues/35768

## Proposed changes:

* Add a toast notification when enabling/disabling cache logging

image

image

* Add a notification manager to fix overlapping notifications.

https://github.com/Automattic/jetpack/assets/11799079/728d752f-6c6f-44a9-87d8-34486a0a490c

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

n/a

## Does this pull request change what data or activity we track or use?

no

## Testing instructions:

* Toggle logging on and off - you should see the notification show up;
* There should be no overlapping notifications.

### CODE REVIEW

**Comment by haqadn** [comment]
The toasts currently overlap each other, and the overlap doesn't depend on which sequence they are dispatched. I think we need a mechanism to handle the toasts central to the app and be able to set position/overlap based on that.

**Comment by thingalon** [comment]
> The toasts currently overlap each other, and the overlap doesn't depend on which sequence they are dispatched. I think we need a mechanism to handle the toasts central to the app and be able to set position/overlap based on that.

What steps do you need to take to see overlapping toasts?

**Comment by haqadn** [comment]
> > The toasts currently overlap each other, and the overlap doesn't depend on which sequence they are dispatched. I think we need a mechanism to handle the toasts central to the app and be able to set position/overlap based on that.
> 
> What steps do you need to take to see overlapping toasts?

Do a combination of:
- Save URL exceptions
- Toggle logging
- Clear cache

**Comment by dilirity** [comment]
@haqadn thanks for reminding me of this! I was aware during development, but totally forgot about it. I wanted to introduce a manager that handles showing the notices to avoid overlapping.

I've added one via https://github.com/Automattic/jetpack/pull/35899/commits/2d7083645fefbf34663b8cbbe9ab9eb2f3c87ced.

Any notice added to the Settings page will be handled by the manager and shown in the bottom right corner.

Here's what it looks like.

https://github.com/Automattic/jetpack/assets/11799079/728d752f-6c6f-44a9-87d8-34486a0a490c

File: projects/plugins/boost/app/assets/src/js/features/notice/context.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+const NoticeContext = createContext< NoticeContextType | undefined >( undefined );
+
+type NoticeProviderProps = {
+	children: ReactNode;
+};
+
+export const NoticeProvider = ( { children }: NoticeProviderProps ) => {
+	const [ notices, setNotices ] = useState< Notice[] >( [] );
+
+	const addOrUpdateNotice = useCallback( ( notice: Notice ) => {
```
Comment (thingalon) [inline projects/plugins/boost/app/assets/src/js/features/notice/context.tsx]: Would `setNotice` be a simpler name that conveys the same meaning?

File: projects/plugins/boost/app/assets/src/js/features/notice/context.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+const NoticeContext = createContext< NoticeContextType | undefined >( undefined );
+
+type NoticeProviderProps = {
+	children: ReactNode;
+};
+
+export const NoticeProvider = ( { children }: NoticeProviderProps ) => {
+	const [ notices, setNotices ] = useState< Notice[] >( [] );
```
Comment (thingalon) [inline projects/plugins/boost/app/assets/src/js/features/notice/context.tsx]: If notices are identified by a string key, wouldn't storing them in an object by key be far simpler?

**Review by thingalon — DISMISSED** [review]
Seems to work well, but I have a couple of minor tidy-ups I'd recommend. Can be in a followup PR.

File: projects/plugins/boost/app/assets/src/js/features/notice/context.tsx (hunk 1 above)
Comment (dilirity) [reply projects/plugins/boost/app/assets/src/js/features/notice/context.tsx]: You're totally right! I've updated this in 8c012c7.

File: projects/plugins/boost/app/assets/src/js/features/notice/context.tsx (hunk 2 above)
Comment (dilirity) [reply projects/plugins/boost/app/assets/src/js/features/notice/context.tsx]: Thanks for the suggestion! Added in e34717d.

**Comment by dilirity** [comment]
@thingalon I've also added an auto-dismiss (5 seconds) to the notices.

**Review by pyronaur — APPROVED** [review]
Really nicely done! 👏 

:shipit:

**Comment by thingalon** [comment]
Cherry-picked to 3.1.0 in b5745abcd0.

**Pull request merged**